rust/src/write.rs

```diff
@@ -233,87 +249,171 @@ impl<'a, W: Write + Seek> Writer<'a, W> {
         }
+        let next_schema_id = self.schemas.len() as u16 + 1;
+        self.write_schema(Schema {
+            id: next_schema_id,
+            name: name.into(),
+            encoding: encoding.into(),
+            data: Cow::Owned(data.into()),
+        })?;
+        Ok(next_schema_id)
+    }
 
-        let next_channel_id = self.channels.len() as u16;
-        assert!(self
-            .channels
-            .insert(chan.clone(), next_channel_id)
-            .is_none());
+    fn write_schema(&mut self, schema: Schema) -> McapResult<()> {
+        self.schemas.insert(
+            SchemaContent {
+                name: Cow::Owned(schema.name.clone()),
+                encoding: Cow::Owned(schema.encoding.clone()),
+                data: Cow::Owned(schema.data.clone().into_owned()),
+            },
+            schema.id,
+        );
         if self.options.use_chunks {
-            self.chunkin_time()?
-                .write_channel(next_channel_id, schema_id, chan)?;
+            self.chunkin_time()?.write_schema(schema)
         } else {
-            write_record(
-                self.finish_chunk()?,
-                &Record::Channel(records::Channel {
-                    id: next_channel_id,
-                    schema_id,
-                    topic: chan.topic.clone(),
-                    message_encoding: chan.message_encoding.clone(),
-                    metadata: chan.metadata.clone(),
-                }),
-            )?;
+            let header = records::SchemaHeader {
+                id: schema.id,
+                name: schema.name,
+                encoding: schema.encoding,
+            };
+            Ok(write_record(
+                &mut self.finish_chunk()?,
+                &Record::Schema {
+                    header,
+                    data: schema.data,
+                },
+            )?)
         }
-        Ok(next_channel_id)
     }
 
-    fn add_schema(&mut self, schema: &Schema<'a>) -> McapResult<u16> {
-        if let Some(id) = self.schemas.get(schema) {
-            return Ok(*id);
+    /// Adds a channel, returning its ID. If a channel with equivalent content was added previously,
+    /// its ID is returned.
+    ///
+    /// Provide a schema_id returned from [`Self::add_schema`], or 0 if the channel has no schema.
+    ///
+    /// Useful with subequent calls to [`write_to_known_channel()`](Self::write_to_known_channel)
+    pub fn add_channel(
+        &mut self,
+        schema_id: u16,
+        topic: &str,
+        message_encoding: &str,
+        metadata: &BTreeMap<String, String>,
+    ) -> McapResult<u16> {
+        if let Some(&id) = self.channels.get_by_left(&ChannelContent {
+            topic: Cow::Borrowed(topic),
+            schema_id,
+            message_encoding: Cow::Borrowed(message_encoding),
+            metadata: Cow::Borrowed(metadata),
+        }) {
+            return Ok(id);
+        }
+        let next_channel_id = self.channels.len() as u16;
```

Comment:
Pedantic: We should probably return an error if there are more than 2^16 channels.